### PR TITLE
var: add TTL and delay timer functionality to new pkg

### PR DESCRIPTION
**Pull request by jrasell**
Targets feature branch.
Related https://github.com/hashicorp/nomad/issues/17449

### CODE REVIEW

File: nomad/lock/delay.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	_ = time.AfterFunc(delay, func() {
+		d.lock.Lock()
+		delete(d.delayTimers, id)
+		d.lock.Unlock()
+	})
+}
+
+// EmitMetrics is a long-running routine used to emit periodic metrics about
+// the Delay.
+func (d *DelayTimer) EmitMetrics(period time.Duration, shutdownCh chan struct{}) {
```
Comment (Juanadelacuesta) [inline nomad/lock/delay.go:63]: Why not use a context instead of a shutdown channel?

File: nomad/lock/delay.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+		case <-timer.C:
+			metrics.SetGauge([]string{"variables", "locks", "delay_timer", "num"}, float32(d.len()))
+		case <-shutdownCh:
+			return
+		}
+	}
+}
+
+// len returns the number of registered delay timers.
+func (d *DelayTimer) len() int {
```
Comment (Juanadelacuesta) [inline nomad/lock/delay.go:79]:
```suggestion
func (d *DelayTimer) timersCount() int {
```

File: nomad/lock/ttl.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	defer t.lock.Unlock()
+
+	if tm := t.ttlTimers[id]; tm != nil {
+		tm.Stop()
+		delete(t.ttlTimers, id)
+	}
+}
+
+// StopAll stops and removes all registered timers.
+func (t *TTLTimer) StopAll() {
```
Comment (Juanadelacuesta) [inline nomad/lock/ttl.go:73]:
```suggestion
func (t *TTLTimer) StopAndRemoveAll() {
```

File: nomad/lock/ttl.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+	if tm := t.ttlTimers[id]; tm != nil {
+		tm.Reset(ttl)
+		return
+	}
+	t.ttlTimers[id] = time.AfterFunc(ttl, afterFn)
+}
+
+// Stop stops the timer with the given ID and removes it from tracking.
+func (t *TTLTimer) Stop(id string) {
```
Comment (Juanadelacuesta) [inline nomad/lock/ttl.go:62]:
```suggestion
func (t *TTLTimer) StopAndRemove(id string) {
```

File: nomad/lock/delay.go (hunk 1 above)
Comment (jrasell) [reply nomad/lock/delay.go:63]: This is to keep consistency with how Nomad does this for other metric emission processes. If we wanted to use a context, we should change all processes that use the shutdownCh to also use this.

File: nomad/lock/delay.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+	// ID, which the caller dictates when adding entries. The lock should be
+	// used for all access.
+	delayTimers map[string]time.Time
+	lock        sync.RWMutex
+}
+
+// NewDelayTimer returns a new delay timer manager.
+func NewDelayTimer() *DelayTimer {
+	return &DelayTimer{
+		delayTimers: make(map[string]time.Time),
```
Comment (pkazmierczak) [inline nomad/lock/delay.go:31]: Wouldn't it be good to initialize the mutex in the constructor, too? I would expect trying to get a lock on an uninitialized mutex to cause a panic.

**Review by pkazmierczak — APPROVED** [review]
LGTM!

**Review by Juanadelacuesta — APPROVED** [review]

**Pull request merged**